contrib/initramfs/hooks/zfs.in

```diff
@@ -3,115 +3,54 @@
-COPY_FILE_LIST="$COPY_FILE_LIST @sysconfdir@/zfs/zfs-functions"
-COPY_FILE_LIST="$COPY_FILE_LIST @sysconfdir@/zfs/vdev_id.conf"
-COPY_FILE_LIST="$COPY_FILE_LIST @udevruledir@/60-zvol.rules"
-COPY_FILE_LIST="$COPY_FILE_LIST @udevruledir@/69-vdev.rules"
-
-# These prerequisites are provided by the base system.
-COPY_EXEC_LIST="$COPY_EXEC_LIST /usr/bin/dirname /bin/hostname /sbin/blkid"
-COPY_EXEC_LIST="$COPY_EXEC_LIST /usr/bin/env"
-COPY_EXEC_LIST="$COPY_EXEC_LIST $(which systemd-ask-password)"
-
-# Explicitly specify all kernel modules because automatic dependency resolution
-# is unreliable on many systems.
-BASE_MODULES="zlib_deflate spl zavl zcommon znvpair zunicode zlua zfs icp"
-CRPT_MODULES="sun-ccm sun-gcm sun-ctr"
-MANUAL_ADD_MODULES_LIST="$BASE_MODULES"
-
-# Generic result code.
-RC=0
+if [ "$1" = "prereqs" ]; then
+	echo "udev"
+	exit
+fi
 
-case $1 in
-prereqs)
-	echo "$PREREQ"
-	exit 0
-	;;
-esac
+. /usr/share/initramfs-tools/hook-functions
 
-for ii in $COPY_EXEC_LIST
-do
-	if [ ! -x "$ii" ]
-	then
-		echo "Error: $ii is not executable."
-		RC=2
-	fi
+for req in "@sbindir@/zpool" "@sbindir@/zfs" "@mounthelperdir@/mount.zfs"; do
+	copy_exec "$req" || {
+		echo "$req not available!" >&2
+		exit 2
+	}
 done
 
-if [ "$RC" -ne 0 ]
-then
-	exit "$RC"
+copy_exec "@sbindir@/zdb"
+copy_exec "@udevdir@/vdev_id"
+copy_exec "@udevdir@/zvol_id"
+if command -v systemd-ask-password > /dev/null; then
+	copy_exec "$(command -v systemd-ask-password)"
 fi
 
-. /usr/share/initramfs-tools/hook-functions
-
-mkdir -p "$DESTDIR/etc/"
-
-# ZDB uses pthreads for some functions, but the library dependency is not
-# automatically detected. The `find` utility and extended `cp` options are
-# used here because libgcc_s.so could be in a subdirectory of /lib for
-# multi-arch installations.
-cp --target-directory="$DESTDIR" --parents $(find /lib/ -type f -name libgcc_s.so.1)
+# We use pthreads, but i-t from buster doesn't automatically
+# copy this indirect dependency: this can be removed when buster finally dies.
+for libgcc in $(find /lib/ -type f -name libgcc_s.so.[1-9]); do
+	copy_exec "$libgcc"
+done
 
-if [ @LIBFETCH_DYNAMIC@ != 0 ]
-then
-	for l in $(find /lib/ -name @LIBFETCH_SONAME@)
-	do
-		copy_exec "$l"
+if [ @LIBFETCH_DYNAMIC@ != 0 ]; then
+	for libfetch in $(find /lib/ -name @LIBFETCH_SONAME@); do
+		copy_exec "$libfetch"
 	done
 fi
 
-for ii in $COPY_EXEC_LIST
-do
-	copy_exec "$ii"
-done
+copy_file config "/etc/hostid"
+copy_file cache  "@sysconfdir@/zfs/zpool.cache"
+copy_file config "@initconfdir@/zfs"
+copy_file config "@sysconfdir@/zfs/zfs-functions"
+copy_file config "@sysconfdir@/zfs/vdev_id.conf"
+copy_file rule   "@udevruledir@/60-zvol.rules"
+copy_file rule   "@udevruledir@/69-vdev.rules"
 
-for ii in $COPY_FILE_LIST
-do
-	dir=$(dirname "$ii")
-	[ -d "$dir" ] && mkdir -p "$DESTDIR/$dir"
-	[ -f "$ii" ] && cp -p "$ii" "$DESTDIR/$ii"
-done
+manual_add_modules zlib_deflate spl zavl zcommon znvpair zunicode zlua zfs icp
```

Comment:
It looks like the `zzstd` module needs to be added here.  I imagine it must have been getting automatically pulled in as a dependency somehow for this to work.

Reply:
I just blindly copied it, but if you wanna touch this, I'd actually say we want to just say "zfs" here: `manual_add_modules()` just asks modprobe for dependencies (i.e. `modprobe --all --set-version="$version" --quiet --show-depends $MODS`, roughly, to paraphrase `hook-functions`, and that's the case as far back as *jessie*): this is to say that it is, *by definition*, correct. If it weren't, the modules wouldn't've loaded into the live system.

I don't know why that comment said that "automatic dependency resolution is unreliable on many systems" but it cannot physically be right – I assume it's folk wisdom from twenty years ago when modular kernels were new and scary.

`dracut-install -m` (as called by `instmods`) does an equivalent thing (but via libkmod and very weirdly, normal day in dracut land) – you will notice that dracut's script installs all the modules explicitly as well, and also doesn't list zzstd (though I'd hesitate to call it "obviously correct" like is the case here).

Reply:
Yup, let's try and do the right thing here by just listing `zfs` and letting depmod do its thing.  Thanks for updating it.